File: bindings/c/src/framework.zig
//! C-compatible framework lifecycle exports.
//! Provides initialization, shutdown, and version functions for FFI.

const std = @import("std");
const errors = @import("errors.zig");

/// Opaque framework handle for C API.
pub const FrameworkHandle = opaque {};

/// Options struct matching C header (abi_options_t).
pub const Options = extern struct {
    enable_ai: bool = true,
    enable_gpu: bool = true,
    enable_database: bool = true,
    enable_network: bool = true,
    enable_web: bool = true,
    enable_profiling: bool = true,
};

/// Version info struct matching C header (abi_version_t).
pub const VersionInfo = extern struct {
    major: c_int,
    minor: c_int,
    patch: c_int,
    full: [*:0]const u8,
};

// Version constants
const VERSION_MAJOR: c_int = 0;
const VERSION_MINOR: c_int = 5;
const VERSION_PATCH: c_int = 0;
const VERSION_STRING: [*:0]const u8 = "0.5.0";

/// Internal framework state stored alongside the handle.
const FrameworkState = struct {
    allocator: std.mem.Allocator,
    options: Options,
    initialized: bool = false,
};

// Global allocator for C bindings
var gpa = std.heap.GeneralPurposeAllocator(.{}){};

// Track active framework to prevent double-init
var active_framework: ?*FrameworkState = null;

/// Initialize framework with defaults.
pub export fn abi_init(out_framework: *?*FrameworkHandle) errors.Error {
    const opts = Options{};
    return abi_init_with_options(&opts, out_framework);
}

/// Initialize framework with options.
pub export fn abi_init_with_options(
    options: *const Options,
    out_framework: *?*FrameworkHandle,
) errors.Error {
    if (active_framework != null) {
        return errors.ALREADY_INITIALIZED;
    }

    const allocator = gpa.allocator();

    const state = allocator.create(FrameworkState) catch {
        return errors.OUT_OF_MEMORY;
    };

    state.* = .{
        .allocator = allocator,
        .options = options.*,
        .initialized = true,
    };

    active_framework = state;
    out_framework.* = @ptrCast(state);
    return errors.OK;
}

/// Shutdown framework.
pub export fn abi_shutdown(handle: ?*FrameworkHandle) void {
    if (handle) |h| {
        const state: *FrameworkState = @ptrCast(@alignCast(h));
        if (state.initialized) {
            state.initialized = false;
            active_framework = null;
            state.allocator.destroy(state);
        }
    }
}

/// Get version string.
pub export fn abi_version() [*:0]const u8 {
    return VERSION_STRING;
}

/// Get detailed version info.
pub export fn abi_version_info(out_version: *VersionInfo) void {
    out_version.* = .{
        .major = VERSION_MAJOR,
        .minor = VERSION_MINOR,
        .patch = VERSION_PATCH,
        .full = VERSION_STRING,
    };
}

/// Check if feature is enabled.
pub export fn abi_is_feature_enabled(
    handle: ?*FrameworkHandle,
    feature: [*:0]const u8,
) bool {
    if (handle) |h| {
        const state: *FrameworkState = @ptrCast(@alignCast(h));
        const feature_str = std.mem.span(feature);

        if (std.mem.eql(u8, feature_str, "ai")) return state.options.enable_ai;
        if (std.mem.eql(u8, feature_str, "gpu")) return state.options.enable_gpu;
        if (std.mem.eql(u8, feature_str, "database")) return state.options.enable_database;
        if (std.mem.eql(u8, feature_str, "network")) return state.options.enable_network;
        if (std.mem.eql(u8, feature_str, "web")) return state.options.enable_web;
        if (std.mem.eql(u8, feature_str, "profiling")) return state.options.enable_profiling;
    }
    return false;
}

test "framework exports" {
    // Reset state for test
    active_framework = null;

    var handle: ?*FrameworkHandle = null;

    // Test version
    const ver = abi_version();
    try std.testing.expectEqualStrings("0.5.0", std.mem.span(ver));

    var info: VersionInfo = undefined;
    abi_version_info(&info);
    try std.testing.expectEqual(@as(c_int, 0), info.major);
    try std.testing.expectEqual(@as(c_int, 5), info.minor);

    // Test init/shutdown
    try std.testing.expectEqual(errors.OK, abi_init(&handle));
    try std.testing.expect(handle != null);

    // Double init should fail
    var handle2: ?*FrameworkHandle = null;
    try std.testing.expectEqual(errors.ALREADY_INITIALIZED, abi_init(&handle2));

    // Test feature check
    try std.testing.expect(abi_is_feature_enabled(handle, "ai"));
    try std.testing.expect(abi_is_feature_enabled(handle, "gpu"));
    try std.testing.expect(!abi_is_feature_enabled(handle, "unknown"));

    // Test shutdown
    abi_shutdown(handle);
    try std.testing.expectEqual(@as(?*FrameworkState, null), active_framework);

    // Shutdown null should be safe
    abi_shutdown(null);
}

File: bindings/rust/src/framework.rs
//! ABI Framework wrapper.

use crate::error::{check_error, Error, Result};
use crate::ffi::{self, AbiFramework, AbiOptions, AbiVersion};
use std::ffi::{CStr, CString};
use std::ptr;

/// ABI Framework instance.
///
/// This is the main entry point for using the ABI framework.
/// It must be initialized before using other ABI features.
///
/// # Example
///
/// ```no_run
/// use abi::Framework;
///
/// let framework = Framework::new().expect("Failed to initialize");
/// println!("ABI version: {}", framework.version());
/// ```
pub struct Framework {
    handle: AbiFramework,
}

impl Framework {
    /// Initialize the ABI framework with default options.
    pub fn new() -> Result<Self> {
        let mut handle: AbiFramework = ptr::null_mut();
        let err = unsafe { ffi::abi_init(&mut handle) };
        check_error(err)?;

        if handle.is_null() {
            return Err(Error::NullPointer);
        }

        Ok(Self { handle })
    }

    /// Initialize the ABI framework with custom options.
    pub fn with_options(options: &Options) -> Result<Self> {
        let mut handle: AbiFramework = ptr::null_mut();
        let abi_options = options.to_ffi();
        let err = unsafe { ffi::abi_init_with_options(&abi_options, &mut handle) };
        check_error(err)?;

        if handle.is_null() {
            return Err(Error::NullPointer);
        }

        Ok(Self { handle })
    }

    /// Get the framework handle for use with other ABI types.
    pub(crate) fn handle(&self) -> AbiFramework {
        self.handle
    }

    /// Get the ABI version string.
    pub fn version(&self) -> &'static str {
        unsafe {
            let ptr = ffi::abi_version();
            if ptr.is_null() {
                "unknown"
            } else {
                CStr::from_ptr(ptr).to_str().unwrap_or("unknown")
            }
        }
    }

    /// Get detailed version information.
    pub fn version_info(&self) -> VersionInfo {
        let mut ver = AbiVersion {
            major: 0,
            minor: 0,
            patch: 0,
            full: ptr::null(),
        };
        unsafe {
            ffi::abi_version_info(&mut ver);
        }

        VersionInfo {
            major: ver.major,
            minor: ver.minor,
            patch: ver.patch,
            full: unsafe {
                if ver.full.is_null() {
                    "unknown".to_string()
                } else {
                    CStr::from_ptr(ver.full)
                        .to_str()
                        .unwrap_or("unknown")
                        .to_string()
                }
            },
        }
    }

    /// Check if a feature is enabled.
    pub fn is_feature_enabled(&self, feature: &str) -> bool {
        let c_feature = CString::new(feature).unwrap_or_default();
        unsafe { ffi::abi_is_feature_enabled(self.handle, c_feature.as_ptr()) }
    }
}

impl Drop for Framework {
    fn drop(&mut self) {
        if !self.handle.is_null() {
            unsafe {
                ffi::abi_shutdown(self.handle);
            }
        }
    }
}

// Framework is not thread-safe by default
// Wrap in Arc<Mutex<Framework>> for multi-threaded use
unsafe impl Send for Framework {}

/// Framework initialization options.
#[derive(Debug, Clone)]
pub struct Options {
    /// Enable AI features (agents, LLM, embeddings).
    pub enable_ai: bool,
    /// Enable GPU acceleration.
    pub enable_gpu: bool,
    /// Enable vector database.
    pub enable_database: bool,
    /// Enable distributed networking.
    pub enable_network: bool,
    /// Enable web/HTTP utilities.
    pub enable_web: bool,
    /// Enable performance profiling.
    pub enable_profiling: bool,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            enable_ai: true,
            enable_gpu: true,
            enable_database: true,
            enable_network: true,
            enable_web: true,
            enable_profiling: true,
        }
    }
}

impl Options {
    /// Create a minimal options with only core features.
    pub fn minimal() -> Self {
        Self {
            enable_ai: false,
            enable_gpu: false,
            enable_database: false,
            enable_network: false,
            enable_web: false,
            enable_profiling: false,
        }
    }

    fn to_ffi(&self) -> AbiOptions {
        AbiOptions {
            enable_ai: self.enable_ai,
            enable_gpu: self.enable_gpu,
            enable_database: self.enable_database,
            enable_network: self.enable_network,
            enable_web: self.enable_web,
            enable_profiling: self.enable_profiling,
        }
    }
}

/// Detailed version information.
#[derive(Debug, Clone)]
pub struct VersionInfo {
    /// Major version number.
    pub major: i32,
    /// Minor version number.
    pub minor: i32,
    /// Patch version number.
    pub patch: i32,
    /// Full version string.
    pub full: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_framework_init() {
        // Skip if library not available
        let result = Framework::new();
        if result.is_err() {
            eprintln!("Skipping test: library not available");
            return;
        }

        let framework = result.unwrap();
        let version = framework.version();
        assert!(!version.is_empty());
    }
}
